schnorr_signature_mvp

VictoriaGrasshopper opened the pull request:
First version of Schnorr signatures library (including MuSig and simple signatures protocol)

Reviewers: @Barsik-sus, @Sakapoi, @TheBestTvarynka, @n-romaniv

VictoriaGrasshopper committed:
init commit

n-romaniv commented on README.md:
I would add a simple usage example, or maybe a link to the examples folder.

n-romaniv commented on src/keys_management.rs:
```diff
@@ -0,0 +1,60 @@
+
+pub struct KeyPair {
+    pub(crate) private_key: Secret<Scalar>,
+    pub public_key: RistrettoPoint,
```
Does this have to be pub?

n-romaniv commented on src/schnorr_musig/musig_math.rs:
```diff
@@ -0,0 +1,361 @@
+// Returns:
+// - Partial signature component si
+pub fn partial_signature(
+    message: &String,
```
It is better to use `&str` or even `AsRef<str>` for ergonomics.

n-romaniv commented on src/schnorr_musig/musig_math.rs:
```diff
@@ -0,0 +1,361 @@
+//
+// Returns:
+// - Final signature s as a Scalar
+pub fn aggregate_s(partial_signatures: &Vec<Scalar>) -> Scalar {
```
There's generally no reason to accept `&Vec<T>`. You can use `AsRef<[T]>` whenever you need an immutable aka shared reference, which is the case here.

TheBestTvarynka commented on src/schnorr_musig/musig_math.rs:
```diff
@@ -0,0 +1,361 @@
+pub fn aggregate_public_keys(public_keys: &Vec<RistrettoPoint>) -> RistrettoPoint {
+    // Ensure there are public keys to aggregate
+    if public_keys.is_empty() {
+        panic!("No public keys provided for aggregation.");
```
do not panic. return an error instead

TheBestTvarynka commented on src/schnorr_musig/musig_math.rs:
```diff
@@ -0,0 +1,361 @@
+pub fn aggregate_nonces(nonces: &Vec<RistrettoPoint>) -> RistrettoPoint {
+    // Ensure there are public keys to aggregate
+    if nonces.is_empty() {
+        panic!("No nonce provided for aggregation.");
```
do not panic

TheBestTvarynka commented on src/schnorr_musig/musig_math.rs:
```diff
@@ -0,0 +1,361 @@
+pub fn verify_signature(
+    cap_x: &RistrettoPoint,
+    cap_r: &RistrettoPoint,
+    message: &String,
```
use `&str` or `impl AsRef<str>`

TheBestTvarynka commented on src/schnorr_musig/musig_math.rs:
```diff
@@ -0,0 +1,361 @@
+    left == right
+}
+
+pub fn test_valid_signature_with_n_parties(n: usize) -> bool {
```
shouldn't it be in the `tests` module?

TheBestTvarynka reviewed (changes requested):
good code. I like it. below you can read some advice (take it or leave it):

* usually, functions that create some values based on the random data accept a random number generator as a parameter. [example](https://github.com/Devolutions/sspi-rs/blob/master/src/utils.rs#L41):
```rust
pub fn generate_random_symmetric_key(cipher: &CipherSuite, rnd: &mut OsRng) -> Vec<u8> {
    todo!()
}
```
* [`zeroize`](https://docs.rs/zeroize/latest/zeroize/) crate. it can help clear the memory when the object dropped

VictoriaGrasshopper committed:
Refactor code based on review feedback:

1. Added basic error handling using String for now; custom error types will be introduced in the future.
2. Implemented verification of the quantity of vector elements (e.g., nonces, public keys, and partial signatures) to ensure they match the number of participants.
3. Improved function privacy by making math functions private to simplify the user interface. (KeyPair stays public, because it is needed for simple signature generation as input, but in future defign would be reconsidered)
4. Enhanced ergonomics by using AsRef as function inputs.
5. Updated example linking to the project.
6. Introduced a randomness source for key and nonce generation functions.
7. Separated tracing as a distinct feature.
8. Other minor changes

Barsik-sus commented on src/keys_management.rs:
```diff
@@ -0,0 +1,57 @@
+}
+
+impl KeyPair {
+    pub fn create<R>(mut rng: R) -> KeyPair
```
I think it would be more convenient to call it `new`, and the following method would be `new_with_private_key`

Barsik-sus commented on src/schnorr_musig/musig_protocol/aggregator.rs:
```diff
@@ -0,0 +1,354 @@
+    ///
+    /// Returns an `Ok` result with an Aggregator in the "CalculatingAggregatedSignature" state,
+    /// or an `Err` result with a description if any of the quantities do not match.
+    pub fn new_skip_the_protocol(
```
What do you think about `builder` instead?

Barsik-sus commented on src/schnorr_musig/musig_protocol/aggregator.rs:
```diff
@@ -0,0 +1,354 @@
+use std::marker::PhantomData;
+
+/// Aggregator struct, which manages the aggregation process.
+pub struct Aggregator<'a, S: AggregationState> {
```
[`Bound behavior, not data`](https://github.com/rust-lang-ua/rust_incubator/tree/master/2_idioms/2_3_bound_impl)
